File: LeetCode/494/main.cpp
class Solution {
public:

    int ans = 0;

    void helper(vector<int>& arr, int ind, int res, int target) {
        
        if(res == target && ind == arr.size()) {
            ans++;
            return;
        } 

        if(ind >= arr.size()) {
            return;
        }

       
        // pick + 
        helper(arr, ind + 1, res + arr[ind], target);

        // pick -
        helper(arr, ind + 1, res - arr[ind], target);

        return;

    }

    int findTargetSumWays(vector<int>& arr, int target) {
        
        helper(arr, 0, 0, target);

        return ans;
    }
};

File: LeetCode/494/main.py
class Solution:
    def findTargetSumWays(self, arr: List[int], target: int) -> int:
        def helper(pos, res):
            key = (pos, res)

            if key not in mp:
                if pos == len(arr):
                    return 1 if res == target else 0
                else:
                    # pick +
                    a = helper(pos + 1, res + arr[pos])

                    # pick -
                    b = helper(pos + 1, res - arr[pos])

                    mp[key] = a + b

            return mp[key]

        mp = defaultdict(int)
        ans = helper(0, 0)

        print(ans)

        return ans
